File: .codex/hooks/subagent_exec_guard.py
#!/usr/bin/env python3
"""Deny container and orchestration commands inside spawned subagents."""

from __future__ import annotations

import json
import os
import sys

from hook_response import deny
from shell_command import (
    SHELLS,
    after_options,
    command_substitutions,
    shell_payload,
    shell_segments,
    strip_leading_shell_prefix,
)

# Only the root agent owns container-level validation. A spawned role reports
# the need instead of reaching for the host runtime itself.
CONTAINER_EXECUTABLES = frozenset(
    {
        "docker",
        "docker-compose",
        "helm",
        "k3d",
        "kind",
        "kubectl",
        "minikube",
        "nerdctl",
        "podman",
        "podman-compose",
    }
)
# Wrappers that run their remaining arguments as the real command.
TRANSPARENT_WRAPPERS = frozenset(
    {
        "command",
        "doas",
        "env",
        "nice",
        "nohup",
        "rtk",
        "stdbuf",
        "sudo",
        "time",
        "timeout",
        "xargs",
    }
)
WRAPPER_OPTIONS_WITH_VALUES = frozenset(
    {
        "-C",
        "-g",
        "-n",
        "-p",
        "-s",
        "-u",
        "--chdir",
        "--group",
        "--kill-after",
        "--prompt",
        "--signal",
        "--user",
    }
)
# Wrappers that consume a leading operand before the real command.
WRAPPER_LEADING_OPERANDS = {"timeout": 1}
MAX_DEPTH = 4
DENIAL_REASON = (
    "Container and orchestration commands are denied inside a spawned agent. "
    "The root agent owns container-level and external-service validation, "
    "because it is the only agent that can escalate to the user. Record the "
    "exact command, why the assigned behavior needs it, and what a passing run "
    "would prove in your result artifact, then report that blocker by returning "
    '`"status":"blocked"` in your manifest. Focused checks that run in-process, '
    "such as the project's unit and package test commands, remain available."
)


def _denied_segment(tokens: list[str], depth: int) -> bool:
    if depth > MAX_DEPTH:
        return False
    remaining = strip_leading_shell_prefix(tokens)
    if not remaining:
        return False

    executable = os.path.basename(remaining[0])
    arguments = remaining[1:]

    if executable in CONTAINER_EXECUTABLES:
        return True

    if executable in SHELLS:
        payload = shell_payload(arguments)
        return payload is not None and has_container_command(payload, depth + 1)

    if executable in TRANSPARENT_WRAPPERS:
        wrapped = after_options(
            arguments,
            options_with_values=WRAPPER_OPTIONS_WITH_VALUES,
        )
        wrapped = wrapped[WRAPPER_LEADING_OPERANDS.get(executable, 0) :]
        return bool(wrapped) and _denied_segment(wrapped, depth + 1)

    return False


def has_container_command(command: object, depth: int = 0) -> bool:
    """Return whether a shell command invokes a container or orchestration tool."""
    if not isinstance(command, str) or not command.strip() or depth > MAX_DEPTH:
        return False
    if any(
        has_container_command(substitution, depth + 1)
        for substitution in command_substitutions(command)
    ):
        return True
    return any(
        _denied_segment(segment, depth) for segment in shell_segments(command)
    )


def in_spawned_agent(event: dict[str, object]) -> bool:
    """Return whether the event comes from a spawned agent rather than the root.

    Codex populates `agent_type` from the spawned thread's role and omits it for
    the root session, so its presence is the role boundary this guard enforces.
    """
    agent_type = event.get("agent_type")
    return isinstance(agent_type, str) and bool(agent_type.strip())


def response_for(event: object) -> dict[str, object] | None:
    if not isinstance(event, dict) or not in_spawned_agent(event):
        return None
    tool_input = event.get("tool_input")
    if not isinstance(tool_input, dict):
        return None
    if not has_container_command(tool_input.get("command")):
        return None
    return deny(DENIAL_REASON)


def main() -> int:
    try:
        event = json.load(sys.stdin)
    except (json.JSONDecodeError, UnicodeDecodeError):
        return 0
    response = response_for(event)
    if response is not None:
        print(json.dumps(response))
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: .codex/hooks/subagent_exec_guard.ts
import * as path from "path";
import { deny } from "./lib/hook_response.ts";
import { handleVersion, isRecord, readEvent, runMain, writeJson } from "./lib/hook_runtime.ts";
import {
  SHELLS,
  afterOptions,
  commandSubstitutions,
  shellPayload,
  shellSegments,
  stripLeadingShellPrefix,
} from "./lib/shell_command.ts";

export const VERSION = "1.1.1";

const CONTAINER_EXECUTABLES = new Set([
  "docker",
  "docker-compose",
  "helm",
  "k3d",
  "kind",
  "kubectl",
  "minikube",
  "nerdctl",
  "podman",
  "podman-compose",
]);
// Option arity belongs to the wrapper: sudo -n/-s are flags, unlike
// nice -n and timeout -s. Consuming a flag's next word could hide the CLI.
const WRAPPER_VALUE_OPTIONS: Record<string, string> = {
  command: "",
  doas: "-C -u",
  env: "-u --unset -C --chdir",
  nice: "-n --adjustment",
  nohup: "",
  rtk: "",
  stdbuf: "-i -o -e --input --output --error",
  sudo: "-C -D -g -h -p -R -r -T -t -u --close-from --chdir --group --host --prompt --chroot --role --command-timeout --type --user",
  time: "-f -o --format --output",
  timeout: "-k -s --kill-after --signal",
  xargs: "-a -d -E -I -L -n -P -s --arg-file --delimiter --eof --replace --max-lines --max-args --max-procs --max-chars",
};
const MAX_DEPTH = 4;
export const DENIAL_REASON =
  "Container and orchestration mutations, process control, and commands not confidently " +
  "recognized as read-only inspection belong to the root agent. Read-only inspection " +
  "such as docker logs, docker ps, and kubectl get remains available. Report the exact " +
  "blocked command, why it is needed, and what a passing run would prove in your assigned " +
  "result format. Ordinary in-process checks remain available.";

// Recognize an intentionally bounded CLI subset. Unknown options are not skipped:
// their values could otherwise be mistaken for a read-only command.
function inspectionWords(args: string[], values: string, flags: string, stopAtCommand = false): string[] | undefined {
  const valueOptions = new Set(values.split(" "));
  const flagOptions = new Set(flags.split(" "));
  const words: string[] = [];
  for (let i = 0; i < args.length; i += 1) {
    const arg = args[i];
    if (arg === "--") {
      return [...words, ...args.slice(i + 1)];
    }
    if (!arg.startsWith("-")) {
      if (stopAtCommand) return args.slice(i);
      words.push(arg);
      continue;
    }
    const option = arg.split("=", 1)[0];
    if (valueOptions.has(option)) {
      if (!arg.includes("=")) {
        i += 1;
        if (i >= args.length || args[i].startsWith("-")) return undefined;
      }
    } else if (flagOptions.has(option)) {
      if (arg.includes("=") && !["true", "false"].includes(arg.slice(option.length + 1))) {
        return undefined;
      }
    } else {
      return undefined;
    }
  }
  return words;
}

function isReadOnlyInspection(executable: string, args: string[]): boolean {
  let words: string[] | undefined;
  if (["docker", "podman", "nerdctl"].includes(executable)) {
    const globals = "--context --host -H --namespace";
    const commandArgs = inspectionWords(args, globals, "", true);
    if (commandArgs === undefined || commandArgs.length === 0) return false;
    const followingLogs = commandArgs[0] === "logs";
    words = inspectionWords(commandArgs,
      globals + " --format --filter --tail --since --until --last -n" + (followingLogs ? "" : " -f"),
      "--all -a --quiet -q --no-trunc --size -s --follow --timestamps -t --details --no-stream --no-reset --latest" + (followingLogs ? " -f" : ""));
    if (words === undefined || words.length === 0) return false;
    if (["container", "image", "network", "volume"].includes(words[0])) {
      return words.length > 1 && ["ls", "inspect"].includes(words[1]);
    }
    return ["ps", "logs", "inspect", "images", "stats", "top", "version", "info"].includes(words[0]);
  }
  if (executable === "kubectl") {
    words = inspectionWords(args,
      "--context --cluster --user --kubeconfig --namespace -n --server -s --request-timeout --output -o --selector -l --field-selector --container -c --tail --since --since-time --limit-bytes --sort-by --template --chunk-size",
      "--all-namespaces -A --watch -w --watch-only --show-labels --no-headers --ignore-not-found --all-containers --follow -f --previous -p --timestamps --prefix --containers --use-protocol-buffers");
    return words !== undefined && words.length > 0 &&
      ["get", "describe", "logs", "top", "version", "api-resources", "api-versions"].includes(words[0]);
  }
  // Other orchestration CLIs remain root-owned until their command/options
  // contract is explicitly covered here and by the focused executable tests.
  return false;
}

function deniedSegment(tokens: string[], depth: number): boolean {
  if (depth > MAX_DEPTH) {
    return false;
  }
  let remaining = stripLeadingShellPrefix(tokens);
  // Unwrapping always consumes a word, so a finite chain needs no recursion
  // cap that could turn a still-visible mutation into an allow.
  while (remaining.length > 0) {
    const executable = path.basename(remaining[0]);
    const arguments_ = remaining.slice(1);
    if (CONTAINER_EXECUTABLES.has(executable)) {
      return !isReadOnlyInspection(executable, arguments_);
    }
    if (SHELLS.has(executable)) {
      const payload = shellPayload(arguments_);
      return payload !== undefined && hasContainerCommand(payload, depth + 1);
    }
    if (!(executable in WRAPPER_VALUE_OPTIONS)) return false;
    let wrapped = afterOptions(arguments_, new Set(WRAPPER_VALUE_OPTIONS[executable].split(" ")));
    wrapped = wrapped.slice(executable === "timeout" ? 1 : 0);
    remaining = stripLeadingShellPrefix(wrapped);
  }
  return false;
}

export function hasContainerCommand(command: unknown, depth = 0): boolean {
  if (typeof command !== "string" || command.trim().length === 0 || depth > MAX_DEPTH) {
    return false;
  }
  if (commandSubstitutions(command).some((substitution) => hasContainerCommand(substitution, depth + 1))) {
    return true;
  }
  return shellSegments(command).some((segment) => deniedSegment(segment, depth));
}

export function inSpawnedAgent(event: Record<string, unknown>): boolean {
  const agentType = event.agent_type;
  return typeof agentType === "string" && agentType.trim().length > 0;
}

export function responseFor(event: unknown): Record<string, unknown> | undefined {
  if (!isRecord(event) || !inSpawnedAgent(event)) {
    return undefined;
  }
  const toolInput = event.tool_input;
  if (!isRecord(toolInput)) {
    return undefined;
  }
  if (!hasContainerCommand(toolInput.command)) {
    return undefined;
  }
  return deny(DENIAL_REASON);
}

function main(): number {
  if (handleVersion(VERSION)) {
    return 0;
  }
  const event = readEvent();
  const response = responseFor(event);
  if (response !== undefined) {
    writeJson(response);
  }
  return 0;
}

runMain("subagent_exec_guard", main);
